File: packages/wire/padding_vector_test.go
package wire

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"testing"
)

type PaddingVector struct {
	Name          string      `json:"name"`
	PlaintextHex  string      `json:"plaintext_hex"`
	UnpaddedLen   int         `json:"unpadded_len"`
	BucketSize    int         `json:"bucket_size"`
	PaddedHex     string      `json:"padded_hex"`
	KeyHex        string      `json:"key_hex"`
	SaltHex       string      `json:"salt_hex"`
	Counter       uint64      `json:"counter"`
	Header        FrameHeader `json:"header"`
	SealedHex     string      `json:"sealed_hex"`
	Valid         bool        `json:"valid"`
	ExpectedError string      `json:"expected_error,omitempty"`
}

func loadPaddingVectors(t *testing.T) []PaddingVector {
	t.Helper()
	data, err := os.ReadFile(filepath.Join("testdata", "padding-vectors.json"))
	if err != nil {
		t.Fatalf("reading padding-vectors.json: %v", err)
	}
	var vectors []PaddingVector
	if err := json.Unmarshal(data, &vectors); err != nil {
		t.Fatalf("parsing padding-vectors.json: %v", err)
	}
	return vectors
}

func TestPaddingVectors(t *testing.T) {
	vectors := loadPaddingVectors(t)
	for _, v := range vectors {
		t.Run(v.Name, func(t *testing.T) {
			plaintext, _ := hex.DecodeString(v.PlaintextHex)
			key, _ := hex.DecodeString(v.KeyHex)
			salt, _ := hex.DecodeString(v.SaltHex)
			dir := DirectionalKey{Key: key, Salt: salt}

			if !v.Valid {
				// Negative vector
				if v.SealedHex != "" {
					sealed, _ := hex.DecodeString(v.SealedHex)
					_, err := Open(dir, v.Counter, sealed)
					if err == nil {
						t.Errorf("expected error opening invalid vector %s, got nil", v.Name)
					}
				}
				if v.PaddedHex != "" {
					padded, _ := hex.DecodeString(v.PaddedHex)
					_, err := UnpadPayload(padded)
					if err == nil {
						t.Errorf("expected error unpadding invalid vector %s, got nil", v.Name)
					}
				}
				return
			}

			// Valid vector: verify PadBucketSize
			if bucket := PadBucketSize(len(plaintext)); bucket != v.BucketSize {
				t.Fatalf("PadBucketSize(%d) = %d, want %d", len(plaintext), bucket, v.BucketSize)
			}

			// Verify PadPayload
			padded, err := PadPayload(plaintext)
			if err != nil {
				t.Fatalf("PadPayload: %v", err)
			}
			if hex.EncodeToString(padded) != v.PaddedHex {
				t.Errorf("padded hex mismatch:\ngot:  %s\nwant: %s", hex.EncodeToString(padded), v.PaddedHex)
			}

			// Verify UnpadPayload
			unpadded, err := UnpadPayload(padded)
			if err != nil {
				t.Fatalf("UnpadPayload: %v", err)
			}
			if !bytes.Equal(unpadded, plaintext) {
				t.Errorf("unpadded mismatch: got %x, want %x", unpadded, plaintext)
			}

			// Verify SealPadded
			h := FrameHeaderInput{
				Version:  v.Header.Version,
				Type:     v.Header.Type,
				Flags:    v.Header.Flags &^ FrameFlagPadded,
				FileIdx:  v.Header.FileIdx,
				BlockIdx: v.Header.BlockIdx,
				FrameOff: v.Header.FrameOff,
			}
			sealed, err := SealPadded(dir, v.Counter, h, plaintext)
			if err != nil {
				t.Fatalf("SealPadded: %v", err)
			}
			if hex.EncodeToString(sealed) != v.SealedHex {
				t.Errorf("sealed hex mismatch:\ngot:  %s\nwant: %s", hex.EncodeToString(sealed), v.SealedHex)
			}

			// Verify Open
			opened, err := Open(dir, v.Counter, sealed)
			if err != nil {
				t.Fatalf("Open: %v", err)
			}
			if !bytes.Equal(opened.Plaintext, plaintext) {
				t.Errorf("opened plaintext mismatch: got %x, want %x", opened.Plaintext, plaintext)
			}
			if opened.Header.Flags&FrameFlagPadded == 0 {
				t.Errorf("expected FrameFlagPadded in opened header")
			}
		})
	}
}

File: packages/protocol/src/padding-vector.test.ts
import { describe, it, expect } from 'vitest';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { hexToBytes, bytesToHex } from './bytes.js';
import { padBucketSize, padPayload, unpadPayload } from './frame.js';
import { open, sealPadded } from './aead.js';
import { FRAME_FLAG_PADDED } from './constants.js';

interface PaddingVector {
  name: string;
  plaintext_hex: string;
  unpadded_len: number;
  bucket_size: number;
  padded_hex: string;
  key_hex: string;
  salt_hex: string;
  counter: number;
  header: {
    Version: number;
    Type: number;
    Flags: number;
    FileIdx: number;
    BlockIdx: number;
    FrameOff: number;
    Len: number;
  };
  sealed_hex: string;
  valid: boolean;
  expected_error?: string;
}

function loadPaddingVectors(): PaddingVector[] {
  const url = new URL('../../wire/testdata/padding-vectors.json', import.meta.url);
  return JSON.parse(readFileSync(fileURLToPath(url), 'utf8')) as PaddingVector[];
}

describe('padding vectors cross-feed', () => {
  const vectors = loadPaddingVectors();

  for (const v of vectors) {
    it(`verifies vector: ${v.name}`, async () => {
      const plaintext = hexToBytes(v.plaintext_hex);
      const key = hexToBytes(v.key_hex);
      const salt = hexToBytes(v.salt_hex);
      const dir = { key, salt };

      if (!v.valid) {
        if (v.padded_hex) {
          const padded = hexToBytes(v.padded_hex);
          expect(() => unpadPayload(padded)).toThrow();
        }
        if (v.sealed_hex) {
          const sealed = hexToBytes(v.sealed_hex);
          await expect(open(dir, v.counter, sealed)).rejects.toThrow();
        }
        return;
      }

      // 1. Bucket size computation
      expect(padBucketSize(plaintext.length)).toBe(v.bucket_size);

      // 2. Padding encoding
      const padded = padPayload(plaintext);
      expect(bytesToHex(padded)).toBe(v.padded_hex);

      // 3. Padding decoding
      const unpadded = unpadPayload(padded);
      expect(bytesToHex(unpadded)).toBe(v.plaintext_hex);

      // 4. Sealing
      const headerInput = {
        version: v.header.Version,
        type: v.header.Type,
        flags: v.header.Flags & ~FRAME_FLAG_PADDED,
        fileIdx: v.header.FileIdx,
        blockIdx: v.header.BlockIdx,
        frameOff: v.header.FrameOff,
      };
      const sealed = await sealPadded(dir, v.counter, headerInput, plaintext);
      expect(bytesToHex(sealed)).toBe(v.sealed_hex);

      // 5. Opening
      const opened = await open(dir, v.counter, sealed);
      expect(bytesToHex(opened.plaintext)).toBe(v.plaintext_hex);
      expect(opened.header.flags & FRAME_FLAG_PADDED).not.toBe(0);
    });
  }
});
